paddle/fluid/operators/merge_selected_var_op.cc

```diff
@@ -0,0 +1,160 @@
+/* Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/fluid/framework/op_registry.h"
+#include "paddle/fluid/memory/memcpy.h"
+
+namespace paddle {
+namespace operators {
+
+class MergeSelectedVarOp : public framework::OperatorBase {
+ public:
+  MergeSelectedVarOp(const std::string &type,
+                     const framework::VariableNameMap &inputs,
+                     const framework::VariableNameMap &outputs,
+                     const framework::AttributeMap &attrs)
+      : OperatorBase(type, inputs, outputs, attrs) {}
+
+ private:
+  void RunImpl(const framework::Scope &scope,
+               const platform::Place &dev_place) const override {
+    platform::DeviceContextPool &pool = platform::DeviceContextPool::Instance();
+    auto &dev_ctx = *pool.Get(dev_place);
+
+    auto &mask = scope.FindVar(Input("Mask"))->Get<framework::LoDTensor>();
+    size_t output_branch = static_cast<size_t>(GetOutputBranch(mask, dev_ctx));
+
+    const std::vector<std::string> &x_names = Inputs("X");
+    PADDLE_ENFORCE_LT(output_branch, x_names.size(),
+                      "Selected branch number is greater than actual branch "
+                      "num in MergeSelectedVarOp");
+
+    const framework::Variable *selected_x =
+        scope.FindVar(x_names[output_branch]);
+
+    if (selected_x->IsType<framework::LoDTensor>()) {
+      auto &to_copy = selected_x->Get<framework::LoDTensor>();
+      auto *out =
+          scope.FindVar(Output("Out"))->GetMutable<framework::LoDTensor>();
+      framework::TensorCopy(to_copy, to_copy.place(), dev_ctx, out);
+      out->set_lod(to_copy.lod());
+    } else if (selected_x->IsType<framework::SelectedRows>()) {
```

Comment:
I think that it is not necessary to support `SelectedRows`.

Reply:
Deleted